File: impl/python/src/img_drv/nix/normalize.py
"""What :mod:`img_drv.nix.emit` forgets, named precisely.

``emit`` and ``parse`` are the two arrows between EXPR and source text. Their
law is a RETRACTION, ``parse(emit(e)) == e``, but it does not hold on the nose,
and the three things it holds only up to are worth naming rather than hiding,
because each one says something about Nix.

Literal CHUNKING carries no meaning
-----------------------------------

Nix's lexer splits a string into chunks at boundaries that depend on which
ESCAPES were used, not on the value: an escaped dollar between two words gives
three parts, and the same characters written plainly give one.
``nix-instantiate --parse`` shows that difference, printing
``("a" + "$" + "b")`` for the first and ``"a$b"`` for the second, and Nix
EVALUATES both identically.

So the debug form our differential oracle compares against is FINER than
semantic equality. That is what we want for testing a parser, since it pins
more; it means the round-trip law has to be stated in the quotient, because
``emit`` writes the characters and cannot write the chunking.

An indented string stops existing at parse time
------------------------------------------------

Nix has no indented-string node: ``''a''`` is an ``ExprString`` once the dedent
has run, and the indentation is gone for good. :class:`~img_drv.nix.ast.IndStr`
is our invention and holds nothing that could write the original back.

A URI literal is a string
--------------------------

``x:x`` is a URI to the LEXER, which is why it is not a lambda, and ``--parse``
prints it as ``"x:x"``. Nix keeps no URI node either.

All three quotients are semantic no-ops, and each names a node WE invented that
Nix does not keep. Normalising by them is the honest statement of the law, not a
way of making a failing test pass; that all three are our own inventions is
itself the finding.
"""

from __future__ import annotations

from . import ast

__all__ = ["expr"]


def _parts(parts: tuple[ast.Part, ...]) -> tuple[ast.Part, ...]:
    out: list[ast.Part] = []
    for p in parts:
        if isinstance(p, ast.Anti):
            out.append(ast.Anti(expr(p.expr)))
            continue
        if out and isinstance(out[-1], ast.Lit):
            out[-1] = ast.Lit(out[-1].text + p.text)
        else:
            out.append(p)
    return tuple(out)


def _attr(a: ast.Attr) -> ast.Attr:
    if isinstance(a, ast.Id):
        return a
    if isinstance(a, ast.StrAttr):
        return ast.StrAttr(_parts(a.parts))
    return ast.DynAttr(expr(a.expr))


def _path(path: ast.AttrPath) -> ast.AttrPath:
    return tuple(_attr(a) for a in path)


def _binding(b: ast.Binding) -> ast.Binding:
    if isinstance(b, ast.Bind):
        return ast.Bind(_path(b.path), expr(b.value))
    source = None if b.source is None else expr(b.source)
    return ast.Inherit(source, _path(b.attrs))


def _pattern(p: ast.Pattern) -> ast.Pattern:
    if isinstance(p, ast.Pvar):
        return p
    formals = tuple(
        (name, None if d is None else expr(d)) for name, d in p.formals
    )
    return ast.Pset(formals, p.ellipsis, p.alias)


def expr(e: ast.Expr) -> ast.Expr:
    """The normal form: literals joined, indented strings and URIs folded."""
    match e:
        case ast.Str():
            return ast.Str(_parts(e.parts))
        # Nix keeps no indented-string node past parsing, so neither does the
        # normal form.
        case ast.IndStr():
            return ast.Str(_parts(e.parts))
        case ast.PathInterp():
            return ast.PathInterp(_parts(e.parts))
        # Nix keeps no URI node either: `x:x` parses to a string.
        case ast.Uri():
            return ast.Str((ast.Lit(e.text),))
        case (
            ast.Int() | ast.Float() | ast.Path() | ast.SearchPath() | ast.Var()
        ):
            return e
        case ast.Lambda():
            return ast.Lambda(_pattern(e.pattern), expr(e.body))
        case ast.Apply():
            return ast.Apply(expr(e.func), expr(e.arg))
        case ast.Select():
            default = None if e.default is None else expr(e.default)
            return ast.Select(expr(e.expr), _path(e.path), default)
        case ast.HasAttr():
            return ast.HasAttr(expr(e.expr), _path(e.path))
        case ast.List():
            return ast.List(tuple(expr(i) for i in e.items))
        case ast.AttrSet():
            binds = tuple(_binding(b) for b in e.binds)
            return ast.AttrSet(binds, e.recursive)
        case ast.Let():
            return ast.Let(tuple(_binding(b) for b in e.binds), expr(e.body))
        case ast.With():
            return ast.With(expr(e.scope), expr(e.body))
        case ast.Assert():
            return ast.Assert(expr(e.condition), expr(e.body))
        case ast.If():
            return ast.If(expr(e.condition), expr(e.then), expr(e.otherwise))
        case ast.Op():
            return ast.Op(e.op, expr(e.left), expr(e.right))
        case ast.Not():
            return ast.Not(expr(e.expr))
        case ast.Neg():
            return ast.Neg(expr(e.expr))

File: impl/ocaml/nix/normalize.ml
(** What {!Emit} forgets, named precisely.

    {!Emit} and {!Nix.parse_string} are the two arrows between EXPR and source
    text. Their law is a RETRACTION, [parse (emit e) = e], but it does not hold
    on the nose, and the two things it holds only up to are worth naming rather
    than hiding, because both say something about Nix.

    {1 Literal CHUNKING carries no meaning}

    Nix's lexer splits a string into chunks at boundaries that depend on the
    ESCAPES used, not on the value: an escaped dollar between two words gives
    three parts, and the same characters written plainly give one.
    [nix-instantiate --parse] shows that difference, printing
    [("a" + "$" + "b")] for the first and ["a$b"] for the second, and Nix
    EVALUATES both identically.

    So the debug form our differential oracle compares against is FINER than
    semantic equality. That is exactly what we want for testing a parser, since
    it pins more; it means a round-trip law has to be stated in the quotient,
    because {!Emit} writes the characters and cannot write the chunking.

    {1 An indented string stops existing at parse time}

    Nix has no indented-string node: [''a''] is an [ExprString] once the dedent
    has run, and the indentation is gone for good. {!Ast.Ind_str} is our
    invention, kept only for the multi-part case, and it holds nothing that
    could be used to write the original back. {!Emit} therefore prints one as
    an ordinary quoted string, and re-parsing gives {!Ast.Str}.

    {1 A URI literal is a string}

    Same story once more: [x:x] is a URI to the LEXER, which is why it is not a
    lambda, and [nix-instantiate --parse] prints it as ["x:x"]. Nix keeps no
    URI node either. {!Emit} writes one as a quoted string, which is both safe
    and preferable, since URI literals are deprecated.

    All three quotients are semantic no-ops, and each one names a node WE
    invented that Nix does not keep. Normalising by them is the honest
    statement of the law, not a way of making a failing test pass; the fact
    that all three are our own inventions is itself the finding. *)

open Ast

let rec merge_parts (parts : part list) : part list =
  match parts with
  | Lit a :: Lit b :: rest -> merge_parts (Lit (a ^ b) :: rest)
  | Anti e :: rest -> Anti (expr e) :: merge_parts rest
  | p :: rest -> p :: merge_parts rest
  | [] -> []

and attr (a : attr) : attr =
  match a with
  | Aid x -> Aid x
  | Astr parts -> Astr (merge_parts parts)
  | Adyn e -> Adyn (expr e)

and bind (b : bind) : bind =
  match b with
  | Bind (path, e) -> Bind (List.map attr path, expr e)
  | Inherit (from, attrs) -> Inherit (Option.map expr from, List.map attr attrs)

and pattern (p : pattern) : pattern =
  match p with
  | Pvar x -> Pvar x
  | Pset f ->
      Pset
        {
          f with
          formals = List.map (fun (n, d) -> (n, Option.map expr d)) f.formals;
        }

(** The normal form: adjacent literals joined, indented strings folded into
    ordinary ones. *)
and expr (e : t) : t =
  match e with
  | Str parts -> Str (merge_parts parts)
  (* Nix keeps no indented-string node past parsing, so neither does the
     normal form. *)
  | Ind_str parts -> Str (merge_parts parts)
  | Path_interp parts -> Path_interp (merge_parts parts)
  (* Nix keeps no URI node either: `x:x` parses to a string. *)
  | Uri u -> Str [Lit u]
  | Int _ | Float _ | Path _ | Search_path _ | Var _ -> e
  | Lambda (p, body) -> Lambda (pattern p, expr body)
  | Apply (f, a) -> Apply (expr f, expr a)
  | Select (e, path, d) -> Select (expr e, List.map attr path, Option.map expr d)
  | Has_attr (e, path) -> Has_attr (expr e, List.map attr path)
  | List items -> List (List.map expr items)
  | Attr_set {recursive; binds} ->
      Attr_set {recursive; binds = List.map bind binds}
  | Let (binds, body) -> Let (List.map bind binds, expr body)
  | With (e, body) -> With (expr e, expr body)
  | Assert (c, body) -> Assert (expr c, expr body)
  | If (c, t, f) -> If (expr c, expr t, expr f)
  | Op (o, a, b) -> Op (o, expr a, expr b)
  | Not e -> Not (expr e)
  | Neg e -> Neg (expr e)
